### PR TITLE
SONARJAVA-2168 Members of Spring components should be '@Autowired'

### PR DESCRIPTION
**Pull request by agigleux**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/spring/S3749Check.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    SymbolMetadata symbolMeta = tree.symbol().metadata();
+
+    if (symbolMeta.isAnnotatedWith("org.springframework.stereotype.Controller")
+      || symbolMeta.isAnnotatedWith("org.springframework.stereotype.Service")
+      || symbolMeta.isAnnotatedWith("org.springframework.stereotype.Repository")) {
+      List<Tree> members = tree.members();
+      for (Tree member : members) {
+        if (member.is(Kind.VARIABLE)) {
+          VariableTree var = (VariableTree) member;
+          if (!var.symbol().isStatic() && !var.symbol().metadata().isAnnotatedWith("org.springframework.beans.factory.annotation.Autowired")) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/spring/S3749Check.java]: CRITICAL Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [S3749Check.java](https://github.com/SonarSource/sonar-java/blob/a64db31022fdc6166806f3a24cca8aec284cc795/java-checks/src/main/java/org/sonar/java/checks/spring/S3749Check.java): Add a new line at the end of this file. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00113)

[CRITICAL]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: java-checks/src/main/java/org/sonar/java/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.SymbolMetadata;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S3749")
+public class SpringComponentWithNonAutowiredMembersCheck extends BaseTreeVisitor implements JavaFileScanner {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: You can use a `IssuableSubscriptionVisitor` and register to `Tree.Kind.INTERFACE, Tree.Kind.ENUM, Tree.Kind.ANNOTATION_TYPE,Tree.Kind.CLASS` kinds of nodes (all implements class tree). 
This should simplify a bit the code and avoid to store the context. Look at other checks extending this class.

It nicely raises the question about if you want to raise issues only on Classes or also on enum interfaces and/or annotation_types.

File: java-checks/src/main/java/org/sonar/java/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  @Override
+  public void visitClass(ClassTree tree) {
+    SymbolMetadata symbolMeta = tree.symbol().metadata();
+
+    if (symbolMeta.isAnnotatedWith("org.springframework.stereotype.Controller")
+      || symbolMeta.isAnnotatedWith("org.springframework.stereotype.Service")
+      || symbolMeta.isAnnotatedWith("org.springframework.stereotype.Repository")) {
+      List<Tree> members = tree.members();
+      for (Tree member : members) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: `members.stream.filter(member.is(Kind.VARIABLE)).forEach( ` is a bit nicer IMO but this is minor

File: java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Controller;
+import org.springframework.stereotype.Service;
+import org.springframework.stereotype.Repository;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+public class HelloWorld {
```
Comment (benzonico) [inline java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: Your testcase lacks a class where it is NOT annotated.

File: java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+  private String name = null; // Noncompliant [[sc=18;ec=22]] {{Make this member @Autowired or remove it.}}
+  public String address = null; // Noncompliant [[sc=17;ec=24]] {{Make this member @Autowired or remove it.}}
+  String phone = null; // Noncompliant [[sc=10;ec=15]] {{Make this member @Autowired or remove it.}}
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(HelloWorld.class); // Compliant
+
+  @RequestMapping(value = "/greet", method = GET)
+  public String greet(String greetee) {
```
Comment (benzonico) [inline java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: The content of the method is irrelevant to the testcase and could be safely removed.

File: java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.stereotype.Repository;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+public class HelloWorld {
+
+  private String name = null; // Noncompliant [[sc=18;ec=22]] {{Make this member @Autowired or remove it.}}
+  public String address = null; // Noncompliant [[sc=17;ec=24]] {{Make this member @Autowired or remove it.}}
+  String phone = null; // Noncompliant [[sc=10;ec=15]] {{Make this member @Autowired or remove it.}}
```
Comment (benzonico) [inline java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: There is also the (nonsense I admit, but kinda important for coverage sake and thinking about how to make your code fail) case missing where a static var is annotated.

File: java-checks/src/main/java/org/sonar/java/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java (hunk 3 above)
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: Well infact you could do all the reporting in one big stream on members : 

```
members.stream.filter(member.is(Kind.VARIABLE))
   .map(m-> (VariableTree) m).
   .filter(v-> v.symbol().isStatic())
   .filter(v-> v.symbol().isAnn.... )
   .forEach(v -> reportIssue(v.simpleName, "..."));
```

File: java-checks/src/main/java/org/sonar/java/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java (hunk 2 above)
Comment (agigleux) [inline java-checks/src/main/java/org/sonar/java/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: Done

File: java-checks/src/main/java/org/sonar/java/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java (hunk 3 above)
Comment (agigleux) [inline java-checks/src/main/java/org/sonar/java/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: Done

File: java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java (hunk 5 above)
Comment (agigleux) [inline java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: Done

File: java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java (hunk 6 above)
Comment (agigleux) [inline java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: Done

File: java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java (hunk 4 above)
Comment (agigleux) [inline java-checks/src/test/files/checks/spring/SpringComponentWithNonAutowiredMembersCheck.java]: Done

**Pull request merged**